paasta_tools/contrib/get_containers_and_ips.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python
+import argparse
+import os
+import socket
+
+from paasta_tools.utils import get_docker_client
+
+HAPROXY_STATS_SOCKET = '/var/run/synapse/haproxy.sock'
+
+
+def send_to_haproxy(command):
+    s = socket.socket(socket.AF_UNIX)
+    # 1 seconds should be more than enough of a timeout since HAProxy is local
+    s.settimeout(1)
+    s.connect(HAPROXY_STATS_SOCKET)
+    print("Sending {} to HAProxy stats socket".format(command))
+    s.send((command + '\n').encode())
+    file_handle = s.makefile()
+    response = file_handle.read().splitlines()
+    print("Response: {}".format(response))
+    s.close()
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=(
+            'Script to dump a HAProxy map between container IPs and task IDs.'
+        ),
+    )
+    parser.add_argument(
+        'map_file', nargs='?', default='/var/run/synapse/maps/ip_to_service.map',
+        help='Where to write the output map file',
+    )
+    args = parser.parse_args()
+
+    client = get_docker_client()
+
+    file_exists = os.path.isfile(args.map_file)
+    if file_exists:
+        mode = 'r+'
+    else:
+        mode = 'w'
+
+    with open(args.map_file, mode) as f:
+        if file_exists:
+            prev_lines = [line.strip().split(' ') for line in f.readlines()]
+            prev_ip_to_task_id = {line[0]: line[1] for line in prev_lines}
+        else:
+            prev_ip_to_task_id = {}
+
+        print("Previous file: {}".format(prev_ip_to_task_id))
+
+        new_lines = []
+        ip_addrs = []
+        for container in client.containers():
```

Comment:
This function is pretty nested and a little hard to read.

As an alternative, what if this loop called a function like `extract_taskid_and_ip` that returned task_id and ip.

Then this script would return "just" the mapping.


Then have a different function that is concerned with taking the existing mapping from the file and comparing the two, and sending the differences off to a function that does the send_to_proxy stuff. (which is concerned with add vs set)

update_haproxy_mappings(new_mappings - old_mappings)
delete_haproxy_mappings(old_mappings - new_mappings)